### ISSUE TITLE
Implement parsing for Meta

### ISSUE THREAD

**Issue by Rantanen** [dtolnay/syn#359]
When using `#[proc_macro_attribute]` the function is defined with two `TokenStream` inputs. These correspond to the attribute parameter tokens and the item tokens.

```rust
#[foo(1, 2, 3)]
struct S;
```

Results in `foo(..)` invocation such as:

```rust
#[proc_maco_attribute]
fn foo(
    attr: TokenStream  // (1, 2, 3)
    item: TokenStream  // struct S;
) { ... }
```

The only way I've found to parse the attribute parameters is to construct the original attribute syntax and parse this:

```rust
let attr_text = format!("#[foo{}]", attr.to_string());
Attribute::parse_outer.parse2(attr_text.parse())
```

**Comment by dtolnay** [comment]
I would be on board with a PR to better support parsing arguments in a proc_macro_attribute.

**Comment by dtolnay** [comment]
I believe we are going to want `syn::parse::Parse` impls for `Meta`, `NestedMeta`, `MetaList`, `MetaNameValue`, and then a `type AttributeArgs = Vec<NestedMeta>`. Attribute macros would use:

```rust
use syn::{parse_macro_input, AttributeArgs, ItemFn};

#[proc_macro_attribute]
fn my_attribute(args: TokenStream, input: TokenStream) {
    let args = parse_macro_input!(args as AttributeArgs);
    let input = parse_macro_input!(input as ItemFn); // or whatever item they work on

    /* ... */
}
```

As part of the same change I would like to add a method `Attribute::parse_meta` with signature `fn(&self) -> syn::Result<Meta>` to leverage the error reporting of the new Parse impls and deprecate `Attribute::interpret_meta`.

@carllerche may be interested in working on this.

**Comment by carllerche** [comment]
I'm just starting to look into this.

I'm not quite sure how `type AttributeArgs = Vec<NestedMeta>` would work. There is no `Parse` implementation for `Vec`.

I could see it working if it was `struct AttributeArgs(Vec<NestedMeta>)`.

**Comment by dtolnay** [comment]
I would prefer not to add a Parse impl for Vec\<NestedMeta\> just yet, but we can do what parse_quote already does and use a separate trait that has impls for all Parse types as well as one for Vec\<NestedMeta\>.
https://github.com/dtolnay/syn/blob/0.15.9/src/parse_quote.rs

**Comment by dtolnay** [comment]
Fixed in #513.
